### PR TITLE
refactor: New Picker interactive 

### PR DESCRIPTION
**Pull request by zombieJ**
## 新交互 (New Design)

image

### RangePicker

- `pickerValue` 逻辑问题
  - [x] Field 切换时自动选择对应的 `pickerValue` ant-design/ant-design#26853
  - [x] `defaultPickerValue` 在未填充时，会根据前一个 field 值自动填充 ant-design/ant-design#43411
  - `pickerValue` 支持与 `onPickerValueChange` 合作受控进行。
    - [x] 📖 `pickerValue` 始终有最高优先级，需要文档说明 ant-design/ant-design#34404
    - [x] 保留 https://github.com/ant-design/ant-design/issues/45382
    - [x] 保留 https://github.com/ant-design/ant-design/issues/44905
    - [x] 保留 https://github.com/ant-design/ant-design/issues/45699
    - [x] 使用 `defaultPickerValue` 或者受控 `pickerValue` 解决 ant-design/ant-design#40680
    - [x] 在已经选择完范围后，重新选择 `startDate` 希望 `endDate` 能跟随 `startDate`
      - ant-design/ant-design#36251
      - https://github.com/ant-design/ant-design/issues/27898
    - [x] 受控 `pickerValue` 控制跳转逻辑
      - ant-design/ant-design#31100
      - https://github.com/ant-design/ant-design/issues/15360
    - [x] 总是在 Input 激活时重置 `pickerValue`
      - ant-design/ant-design#22471
      - https://github.com/ant-design/ant-design/issues/21787
    - [x] `showTime.defaultValue` 可以聚合至 `pickerValue` ant-design/ant-design#44724
    - [x] 固定至第一个有效值，用 `showTime.defaultValue` ant-design/ant-design#25081
- 输入框交互
  - [x] 输入任意键不再唤起弹窗
    - https://github.com/ant-design/ant-design/issues/44203
    - https://github.com/ant-design/ant-design/issues/40243
  - [x] 添加 `preserveInvalidOnBlur=false` 保留无效值 ant-design/ant-design#32475
  - [x] 输入框内容正确时就提交，而不是失去焦点丢失输入
    - https://github.com/ant-design/ant-design/issues/31691
    - https://github.com/ant-design/ant-design/issues/39651
    - https://github.com/ant-design/ant-design/issues/30468
    - https://github.com/ant-design/ant-design/issues/31158
  - [x] 简化范围选择交互样式，移除虚线
    - https://github.com/ant-design/ant-design/issues/39462
    - https://github.com/ant-design/ant-design/issues/27855
  - [x] `format` 支持 `align` 做格式匹配
    - ant-design/ant-design#25404
    - https://github.com/ant-design/ant-design/issues/39321
  - [x] 支持 `required` ant-design/ant-design#27616
  - [x] 支持 `ids` ant-design/ant-design#39629
  - [x] `onFocus` & `onBlur` 支持 `info.range` ant-design/ant-design#34261
  - [x] `focus` event 总是会触发 ant-design/ant-design#45647
- 面板相关
  - [x] 📖 `showTime` 支持 `changeOnScroll`，文档需要添加一个居中滚动的例子
    - https://github.com/ant-design/ant-design/issues/29508
    - https://github.com/ant-design/ant-design/issues/27480
  - [x] 修复 `mode` 受控问题 https://github.com/ant-design/ant-design/issues/19835
  - [x] 不再根据起始或结束时间禁用范围 ant-design/ant-design#28083
  - [x]  支持 `needConfirm` 允许更新需要、不需要确认
    - 让所有 Picker 都需要确认
      - ant-design/ant-design#22966
      - ant-design/ant-design#45896
    - 让所有 Picker 都不需要确认
      - ant-design/ant-design21189
      - ant-design/ant-design#45896
  - [x] 选择中途不再触发 `onChange` 事件 https://github.com/ant-design/ant-design/issues/38293
  - [x] 双击确认支持需要点确认按钮的场景 ant-design/ant-design#44295
  - [x] 受控 `open` 总是能正确触发 `onChange` 事件
    - ant-design/ant-design#30525
    - https://github.com/ant-design/ant-design/issues/32134
    - https://github.com/ant-design/ant-design/issues/46152
    - https://github.com/ant-design/ant-design/issues/35239
    - https://github.com/ant-design/ant-design/issues/39200
  - [x] 选择不同的 `startDate` 和 `endDate` 不再禁用面板点击
    - https://github.com/ant-design/ant-design/issues/28083
  - [x] 输入框宽度变化时重新对齐面板 ant-design/ant-design#42775
  - [x] 面板切换时不再抖动 ant-design/ant-design#44177
  - [x] 支持 `millisec`
    - https://github.com/ant-design/ant-design/issues/28591
    - https://github.com/ant-design/ant-design/issues/16466
    - https://github.com/ant-design/ant-design/issues/28075
- 7 天选择实例问题
  - [x] `onCalendarChange` 在值变化时总是触发
    - ant-design/ant-design#45611
    - https://github.com/ant-design/ant-design/issues/43079
  - [x] `onCalendarChange` 在关闭重置时也需要正确触发 ant-design/ant-design#22351
  - [x] 📖 `onSelect` 合并至 `onCalendarChange` (antd 英文文档需要更新) ant-design/ant-design#33972
  - [x] 新的限制日期范围逻辑，`disabledDate` 提供 `info.from` 告知是选择时间
    - https://github.com/ant-design/ant-design/issues/45528
    - https://github.com/ant-design/ant-design/issues/45909
    - https://github.com/ant-design/ant-design/issues/29703
    - https://github.com/ant-design/ant-design/issues/26840
    - https://github.com/ant-design/ant-design/issues/31081
    - https://github.com/ant-design/ant-design/issues/46187
    - https://github.com/ant-design/ant-design/issues/46194
- 自定义展示能力
  - [x] 📖  `components.input` 支持自定义组件 
    - ant-design/ant-design#45870
    - https://github.com/ant-design/ant-design/issues/32864
    - https://github.com/ant-design/ant-design/issues/36429
    - https://github.com/ant-design/ant-design/issues/23276
    - https://github.com/ant-design/ant-design/issues/45918
    - https://github.com/ant-design/ant-design/issues/45463
  - [x] 📖  支持 `components.button`
    - https://github.com/ant-design/ant-design/issues/21525
    - https://github.com/ant-design/ant-design/issues/25926
  - [x] 📖  `panelRender` 替换完整内容 ant-design/ant-design#21635
  - [x] `renderExtraFooter` 不关闭面板
    - ant-design/ant-design#24696
    - https://github.com/ant-design/ant-design/issues/27040
    - https://github.com/ant-design/ant-design/issues/44005
  - [x] 📖 复用 `allowEmpty` 和 `placeholder` 显示至今（更新 Demo） ant-design/ant-design#42338
  - [x] 支持 `components.time` 自定义面板 https://github.com/ant-design/ant-design/issues/29501
  - [x] 添加 `changeOnScroll` 支持 https://github.com/ant-design/ant-design/issues/29508
  - 自定义 formatter 需要影响所有面板
    - [x] 面板单元的 `title` 现在支持 `locale.fieldXXXFormat` ant-design/ant-design#31339
    - [x] 面板所有内容都支持 `locale.xxxCellFormat`
      - https://github.com/ant-design/ant-design/issues/28689
      - https://github.com/ant-design/ant-design/issues/28022
      - https://github.com/ant-design/ant-design/issues/29625
      - https://github.com/ant-design/ant-design/issues/31520
    - [x] 支持 LTS 格式 https://github.com/ant-design/ant-design/issues/27777
- 禁用相关
  - [x] 修复点击时间可以穿透 `disabledDate` 问题 https://github.com/ant-design/ant-design/issues/29023
  - [x] 修复点击日期可以穿透 `disabledTime` 问题 https://github.com/ant-design/ant-design/issues/20280
  - [x] `showTime` 本身就支持 function `disabledTime` https://github.com/ant-design/ant-design/issues/29179
  - [x] `presets` 禁用时点击不再允许触发 `onChange` https://github.com/ant-design/ant-design/issues/18765
  - [x] `showTime.defaultValue` 可以设置禁用的 `disabledTime` 但是不允许提交 ant-design/ant-design#13721
  - [x] `showNow` 也会经过提交拦截 https://github.com/ant-design/ant-design/issues/30568
  - [x] 自动选择可以使用的 `showTime.disabledTime` 时间 ant-design/ant-design#25143
  - [x] 手工输入禁用时间不触发 `onChange` https://github.com/ant-design/ant-design/issues/45489
  - [x] 手工输入不能穿透点击 `OK` 按钮 ant-design/ant-design#45564
  - [x] 添加 `info.type` 信息，告知当前面板是何面板 ant-design/ant-design#39991
  - [x]  添加 `minDate` & `maxDate` 支持，切换标签会阻止更远的点击
    - ant-design/ant-design#22912
    - https://github.com/ant-design/ant-design/issues/37685
    - https://github.com/ant-design/ant-design/issues/46196
- 杂项
  - [x] `allowClear` 事件交还给 `onClick` 事件 https://github.com/ant-design/ant-design/issues/31799
  - [x] RangePicker 也支持 `order`  ant-design/ant-design#33440
  - [x] 支持 `showWeek` ant-design/ant-design#21726
  - [x] 修复 `disabled` 与 `allowEmpty` warning 时机错误的问题 ant-design/ant-design#26286
  - [x] 移除 `mode` 切换特殊逻辑，现在总是 `year` > `month` > `date` ant-design/ant-design#38343
  - 对齐 Picker 行为，现在 `disabled` 下设置 `open` 不再能打开面板
  - 移除废弃属性 `defaultOpenValue`
  - `onKeyDown` 不再提供 `preventDefault` 方法，请使用 event 原生的 `preventDefault` 方法

### Picker

- [x] `needConfirm` ant-design/ant-design#23475
- [x] `meridiem` 也需要滚动 ant-design/ant-design#45691
- [ ] `mode` 切换实现 `date` -> `time` 选择 ant-design/ant-design#27680
- [x] 🔥 支持 `multiple` ant-design/ant-design#4447
- [ ] 单选面板点击 `Now` 不触发 `onChange` 事件 ant-design/ant-design#28811

### 移动端

- https://github.com/ant-design/ant-design/issues/32011
- https://github.com/ant-design/ant-design/issues/26806
- https://github.com/ant-design/ant-design/issues/26758

### Holder

- https://github.com/ant-design/ant-design/issues/30506
- 单独面板渲染能力（内联渲染 PickerPanel，不需要 Selector）
  - [ ] https://github.com/ant-design/ant-design/issues/43519
  - [ ] https://github.com/ant-design/ant-design/issues/40051

### Ant Design Side

- ConfigProvider support `format`
  - ant-design/ant-design#23902
  - https://github.com/ant-design/ant-design/issues/45416

### CODE REVIEW

**Comment by Eva-HYL** [comment]
你好 请问这个问题更新了么

**Comment by Alvin-F** [comment]
DatePicker 在 picker 设置除了默认 date 之外， 再进行设置 disabledDate 不可选今后时间后，打开及会很慢（卡顿效果） 原因大概是： 做对比时间始终根据天来的；
这个望及时修复

File: docs/examples/debug.tsx [hunk 1]
```diff
@@ -145,7 +147,7 @@ export default () => {
           generateConfig={dayjsGenerateConfig}
           locale={zhCN}
           defaultValue={[dayjs()]}
-          multiple
+          multiple={true as boolean}
```
Comment (afc163) [inline docs/examples/debug.tsx:154]: 。。。为啥要这样

Comment (zombieJ) [reply docs/examples/debug.tsx:169]: 测试类型定义用的，`debug.tsx` 里都是我用来测试的。PR完成后会删掉。

**Comment by zombieJ** [comment]
line cov 差 0.85% 在 dom 测量上，发包后先测 antd。然后再来补。

File: src/PickerInput/Popup/index.tsx [hunk 2]
```diff
@@ -0,0 +1,172 @@
...
+  let renderNode = (
+    <div
+      tabIndex={-1}
+      className={classNames(
+        containerPrefixCls,
+        // Used for Today Button style, safe to remove if no need
+        `${prefixCls}-${internalMode}-panel-container`,
+      )}
+      style={{
+        [rtl ? marginRight : marginLeft]: containerOffset,
```
Comment (li-jia-nan) [inline src/PickerInput/Popup/index.tsx:140]: 大佬这里为啥不用逻辑属性？

File: src/PickerInput/SinglePicker.tsx [hunk 3]
```diff
@@ -0,0 +1,663 @@
...
+            onSelectorInvalid(invalid, 0);
+          }}
+        />
+      </PickerTrigger>
+    </PickerContext.Provider>
+  );
+}
+
+const RefPicker = React.forwardRef(Picker) as <DateType extends object = any>(
+  props: PickerProps<DateType> & { ref?: React.Ref<PickerRef> },
```
Comment (li-jia-nan) [inline src/PickerInput/SinglePicker.tsx:656]: 这里的 `{ ref?: React.Ref<PickerRef> }` 是不是可以写成 `React.RefAttributes<PickerRef>`？

File: src/PickerInput/Popup/index.tsx (hunk 2 above)
Comment (zombieJ) [reply src/PickerInput/Popup/index.tsx:140]: 这里是 inline code，用户没办法通过 transformer 降级

File: src/PickerInput/SinglePicker.tsx (hunk 3 above)
Comment (zombieJ) [reply src/PickerInput/SinglePicker.tsx:656]: 可以的~纯粹我不知道有这个，2333

**Comment by afc163** [comment]
README 中没有 components

**Comment by EliazTray** [comment]
`单独面板渲染能力（内联渲染 PickerPanel，不需要 Selector）`
这个目前有排在计划内吗？

**Pull request merged**